src/net/server/handlers/mul.rs

```diff
@@ -0,0 +1,186 @@
+use crate::net::server::MpcServerError;
+use crate::net::{BufferedMessages, RecordHeaders};
+use crate::protocol::{QueryId, Step};
+use async_trait::async_trait;
+use axum::{
+    body::Bytes,
+    extract::{self, FromRequest, RequestParts},
+    http::{Request, StatusCode},
+    middleware::Next,
+    response::{IntoResponse, Response},
+    Extension,
+};
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use tokio::sync::mpsc;
+use tracing::log::debug;
+
+/// Used in the axum handler to extract the `query_id` and `step` from the path of the request
+pub struct Path<S: Step>(QueryId, S);
+#[async_trait]
+impl<B: Send, S: Step> FromRequest<B> for Path<S> {
+    type Rejection = MpcServerError;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let extract::Path((query_id, step)) =
+            extract::Path::<(QueryId, String)>::from_request(req).await?;
+        let step = S::try_from(step).map_err::<serde_json::Error, _>(serde::de::Error::custom)?;
+        Ok(Path(query_id, step))
+    }
+}
+
+/// TODO: value of map may change to different implementation
+pub type GatewayMap<S> = Arc<Mutex<HashMap<(QueryId, S), mpsc::Sender<BufferedMessages<S>>>>>;
+
+/// Injects the appropriate gateway/mesh into the Axum request, so that downstream handlers have
+/// simple access to the correct value
+///
+/// For now, stub out the gateway/mesh logic with just a channel
+/// TODO: if request asks for gateway that does not exist, should it add the gateway, or should
+///       it fail? who is responsible for managing the existence of gateways?
+pub async fn gateway_middleware_fn<B: Send, S: Step>(
```

Comment:
this is a step in the right direction - is there a particular reason to inject it as middleware rather than shared state?
https://docs.rs/axum/latest/axum/struct.Extension.html

I would imagine IPA application gateway to exist for all requests - that way we can add backpressure, so it would need to be present as a struct and the entry point from axum handler to IPA app

Reply:
obsolete code